### PR TITLE
[W10H01] Simple TrialOfTheDreamsTest

### PR DESCRIPTION
**Pull request by MoLaden**
this time including .java right away

### CODE REVIEW

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static pgdp.trials.TrialOfTheDreams.lockPick;
+
+public class TrialOfTheDreamsTest {
+    // Please note: Method lockPick(Function<Byte[], Boolean> lock, List<Byte> key, int maxlen) has to be set to public
```
Comment (felixhauptmann) [inline w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:16]: [Please do not do this. Use reflection instead.](https://github.com/MaximilianAnzinger/pgdp2223-tests/blob/main/CONTRIBUTING.md#dont-depend-on-changing-the-code-of-the-implementation)

**Review by felixhauptmann — CHANGES_REQUESTED** [review]

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    // for this test to work
+
+    @Test
+    @DisplayName("Basic Example")
+    void testBasicExample() {
+        Function<Byte[], Boolean> testFunction = new Function<Byte[], Boolean>() {
```
Comment (julian-hartl) [inline w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:22]: Could you refactor this into a helper method? For example:
```suggestion
        var testFunction = createTestFunction(bytes -> (bytes.length == 2 && bytes[0] == -97 && bytes [1] == 37))
```

**Review by julian-hartl — CHANGES_REQUESTED** [review]
You could maybe consider using [Parameterized Tests](https://www.baeldung.com/parameterized-tests-junit-5)

**Review by MaximilianAnzinger — APPROVED** [review]

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java (hunk 2 above)
Comment (Anatoly03) [reply w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:22]: @MoLaden [These are the Java Docs for Lambdas](https://docs.oracle.com/javase/tutorial/java/javaOO/lambdaexpressions.html).

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    void testWrapperFunction() {
+        Function<byte[], Boolean> testFunction = new Function<byte[], Boolean>() {
+            @Override
+            public Boolean apply(byte[] bytes) {
+                return (bytes.length == 2 && bytes[0] == -97 && bytes [1] == 37);
+            }
+        };
+        assertArrayEquals(new byte[]{-97, 37}, lockPick(testFunction, 3));
+    }
+}
```
Comment (felixhauptmann) [inline w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:59]: Did it for you :)

```suggestion
package pgdp.trials;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

import static org.junit.jupiter.api.Assertions.*;
import static pgdp.trials.TrialOfTheDreams.lockPick;

public class TrialOfTheDreamsTest {

    @SuppressWarnings("unchecked")

    private List<Byte> callPrivateHelperMethod(Function<Byte[], Boolean> key, List<Byte> list, int maxLen) throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
        Method m = TrialOfTheDreams.class.getDeclaredMethod("lockPick", Function.class, List.class, int.class);
        m.setAccessible(true);
        return (List<Byte>) m.invoke(null, key, list, maxLen);
    }

    @Test
    @DisplayName("Basic Example")
    void testBasicExample() throws InvocationTargetException, NoSuchMethodException, IllegalAccessException {
        Function<Byte[], Boolean> testFunction = bytes -> (bytes.length == 2 && bytes[0] == -97 && bytes[1] == 37);

        List<Byte> res = callPrivateHelperMethod(testFunction, new ArrayList<>(), 2);

        assertEquals(2, res.size());
        assertEquals((byte) -97, res.get(0));
        assertEquals((byte) 37, res.get(1));
    }

    @Test
    @DisplayName("Should return null if key is longer then maxlen")
    void testNullIfNotFound() throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
        Function<Byte[], Boolean> testFunction = bytes -> (bytes.length == 5);

        Method m = TrialOfTheDreams.class.getDeclaredMethod("lockPick", Function.class, List.class, int.class);
        m.setAccessible(true);

        List<Byte> res = callPrivateHelperMethod(testFunction, new ArrayList<>(), 2);
        assertNull(res);
    }

    @Test
    @DisplayName("Tests lockPick(Function<byte[], Boolean> lock, int maxlen)")
    void testWrapperFunction() {
        Function<byte[], Boolean> testFunction = bytes -> (bytes.length == 2 && bytes[0] == -97 && bytes[1] == 37);
        assertArrayEquals(new byte[]{-97, 37}, lockPick(testFunction, 3));
    }
}
```

**Review by felixhauptmann — CHANGES_REQUESTED** [review]

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java (hunk 3 above)
Comment (felixhauptmann) [reply w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:59]: Jesus GitHub why don't you collapse this abomination of a diff.

Sorry, next time I'll create a PR again.

**Pull request closed without merging**